File: research/portal-ui/src/lib/chunker.ts
import { get_encoding } from 'tiktoken';

/* ─── Configuration (mirrors chunker.py) ─── */
const MAX_TOKENS = 500;
const OVERLAP = 50;
const ENCODING_MODEL = 'cl100k_base';

/**
 * Token-based text chunker — direct port of the Python TextChunker.
 * Uses tiktoken with cl100k_base encoding and a sliding window approach.
 */
export function chunkText(text: string): string[] {
    const encoder = get_encoding(ENCODING_MODEL);
    const tokens = encoder.encode(text);
    const chunks: string[] = [];
    const step = MAX_TOKENS - OVERLAP;

    for (let i = 0; i < tokens.length; i += step) {
        const chunkTokens = tokens.slice(i, i + MAX_TOKENS);
        const decoded = new TextDecoder().decode(encoder.decode(chunkTokens));
        chunks.push(decoded);
    }

    encoder.free();
    return chunks;
}

export { MAX_TOKENS, OVERLAP };

File: research/scraper-lambda/chunker.py
import tiktoken
import yaml

with open("config.yml", "r") as f:
    config = yaml.safe_load(f)

class TextChunker:
    def __init__(self):
        self.max_tokens = config['chunking']['max_tokens']
        self.overlap = config['chunking']['overlap']
        # Load the specific tokenizer used by OpenAI
        self.tokenizer = tiktoken.get_encoding(config['chunking']['encoding_model'])

    def chunk(self, text):
        tokens = self.tokenizer.encode(text)
        chunks = []
        
        # Slide a window over the tokens to create overlapping chunks
        for i in range(0, len(tokens), self.max_tokens - self.overlap):
            chunk_tokens = tokens[i : i + self.max_tokens]
            chunks.append(self.tokenizer.decode(chunk_tokens))
            
        return chunks